perf: improve performance for tabs changing in HomePage and page loading when switching to HomePage

EurFelux opened the pull request:
<!-- Template from https://github.com/kubevirt/kubevirt/blob/main/.github/PULL_REQUEST_TEMPLATE.md?-->
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/CherryHQ/cherry-studio/blob/main/CONTRIBUTING.md
-->

### What this PR does

由于现在应用常有渲染卡顿的问题，同时渲染速度不是很好优化，计划使用skeleton优化组件渲染过程的用户体验

- 优化TopicsTab重新渲染时的用户体验和响应速度
- 优化Messages重新渲染时的用户体验和响应速度
- 切换到HomePage时，延后输入栏中工具栏的渲染以增加响应速度
- 优化从其他页面切换到HomePage时的性能 

<details>
<summary>stale changes</summary>
<img width="1470" height="866" alt="image" src="https://github.com/user-attachments/assets/527b7ade-15a0-4431-a14e-bdb8a11a22ae" />
</details>

[video demo](https://github.com/user-attachments/assets/17268e79-ac3d-4937-84f3-41bc5f0b66bb)
https://github.com/CherryHQ/cherry-studio/pull/8246#issuecomment-3109436346

在长对话场景下，切换page的响应速度能达到原来的两倍左右。（⬅️开发，➡️生产，同一话题数据）
<img width="1587" height="562" alt="image" src="https://github.com/user-attachments/assets/6df462a8-2826-42c6-b611-f45b31787e32" />

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->

相关issues：#8237 #8217

Fixes #8303 

### Why we need it and why it was done in this way

The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Breaking changes

<!-- optional -->

If this PR introduces breaking changes, please describe the changes and the impact on users.

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature.

### Release note

<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->

```release-note

```

EurFelux committed:
feat: 添加话题列表加载时的骨架屏效果

使用React的useTransition实现平滑过渡，在话题列表排序时显示骨架屏加载状态

EurFelux committed:
feat(消息列表): 添加加载中的骨架屏效果以提升用户体验

使用React的useTransition实现平滑过渡，在消息加载时显示骨架屏
添加MessageSkeleton组件展示加载状态

EurFelux committed:
fix(TopicsTab): 修复重渲染时话题skeleton再次出现的问题

仅在首次渲染时使用startTransition

EurFelux committed:
fix(消息列表): 处理空消息数组时的UI闪烁问题

在获取到有效消息前，空数组多次触发effect导致UI闪烁。添加条件分支处理空数组情况

EurFelux committed:
feat(inputbar): 延后工具栏渲染

beyondkmp commented:
不错，速度提升了，内存占用会增加吗？

0xfullex commented:
skeleton作为加载中的提示，不适合太密铺满

> <img alt="image" width="1470" height="866" src="https://private-user-images.githubusercontent.com/59059173/467614258-a9ea61de-fede-404f-b456-121c67970088.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI4MDEzNjQsIm5iZiI6MTc1MjgwMTA2NCwicGF0aCI6Ii81OTA1OTE3My80Njc2MTQyNTgtYTllYTYxZGUtZmVkZS00MDRmLWI0NTYtMTIxYzY3OTcwMDg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE4VDAxMTEwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg5NmE2MDRmMmQyZjhkNzNjMDM2MmZiZmY3YzAzZDAzNTA2NWY3ZjhjMmIzNTcwZGZjNDY1NDFmNzk0ZWQ4OWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ytFHDUTp88HqvgyXOiFWQZD9040dNGbKGhxfOAhz_3Y">

beyondkmp commented:
测试了，效果不错。内存和cpu都没有涨。very good.

EurFelux committed:
style(ui): 优化骨架屏样式和数量

调整TopicsTab和Messages组件的骨架屏显示效果
- 减少TopicsTab的骨架屏线条数量
- 优化Messages骨架屏的段落宽度和间距
- 移除多余的MessageSkeleton实例

alephpiece commented on src/renderer/src/pages/home/Tabs/TopicsTab.tsx:
```diff
@@ -432,21 +435,41 @@ const Topics: FC<Props> = ({ assistant: _assistant, activeTopic, setActiveTopic
-        return 0
+
+  useEffect(() => {
+    const updateTopics = () => {
+      if (pinTopicsToTop) {
+        setDisplayedTopics(
+          [...assistant.topics].sort((a, b) => {
+            if (a.pinned && !b.pinned) return -1
+            if (!a.pinned && b.pinned) return 1
+            return 0
+          })
+        )
+      } else {
+        setDisplayedTopics(assistant.topics)
+      }
+    }
+
+    if (isFirstRender) {
+      startTransition(() => {
+        updateTopics()
       })
+      setIsFirstRender(false)
+    } else {
+      updateTopics()
     }
-    return assistant.topics
-  }, [assistant.topics, pinTopicsToTop])
+  }, [assistant.topics, isFirstRender, pinTopicsToTop])
+
+  // FIXME: 如果添加其他transition会导致skeleton重新显示
+  if (isPending) {
+    return <TopicsSkeleton></TopicsSkeleton>
+  }
```
topics 现在是虚拟列表了还会占用很多时间吗？应该只会渲染可见的那些？

EurFelux replied:
而且不知为何DraggableVirtualList有多次渲染和提交的过程，实际响应时间大概60~70ms左右

alephpiece replied:
会卡顿吗？开发模式？开着开发工具肯定是会慢的

alephpiece replied:
main分支，开发模式（关闭开发工具）：
https://github.com/user-attachments/assets/001d3bc0-89d1-4619-93a5-317d3a43b05e

skeleton 确实能改善体验，不过能消除闪烁就更好了，可以在话题数量比较多的时候才出现？

EurFelux committed:
fix(消息组件): 修复发送消息时触发skeleton的问题

添加isFirstRender状态变量以区分首次渲染和后续更新
避免因多次触发重渲染导致的UI闪烁问题

0xfullex reviewed (commented):
发送消息后，skeleton和对话消息会切换闪动几次

0xfullex commented:
> ` [7dc4d96](/CherryHQ/cherry-studio/pull/8246/commits/7dc4d961f47ae63ccaaecb6c6f8666114dc06177)`

更新 7dc4d96 后，新建话题，首次发消息，skeleton会闪动一次

EurFelux committed:
fix(消息组件): 优化消息渲染逻辑

移除空消息数组时的处理，避免发送第一条消息时触发skeleton
将isFirstRender加入骨架屏显示条件

EurFelux commented:
[bcfdf73](https://github.com/CherryHQ/cherry-studio/pull/8246/commits/bcfdf732028096da44a4f802ade051ce29f22d72)这个解决方案有问题，我再看看怎么改

0xfullex commented:
现在在短对话间切换，也出现skeleton一闪而过（本身加载速度就很快了，其实不需要skeleton。

有没有可能判断下条数，条数较少的就不显示skeleton了

alephpiece commented:
> 现在在短对话间切换，也出现skeleton一闪而过（本身加载速度就很快了，其实不需要skeleton。
> 
> 有没有可能判断下条数，条数较少的就不显示skeleton了

消息条数也不可靠😂有的消息很长，包含大量 mcp 调用、代码、图片，渲染一个比渲染好几个纯文字的时间还长

EurFelux commented:
给skeleton设置一个最小显示时间？这样不会闪，但总是要显示一小段时间

EurFelux committed:
Revert "fix(消息组件): 优化消息渲染逻辑"

This reverts commit bcfdf732028096da44a4f802ade051ce29f22d72.

EurFelux committed:
fix(消息组件): 修复首次渲染时消息闪烁问题

优化消息加载逻辑，通过添加pendingCount状态跟踪处理次数，避免首次渲染时的UI闪烁

EurFelux committed:
feat(消息页面): 添加骨架屏延迟显示逻辑

添加骨架屏延迟显示功能，当加载时间小于SKELETON_DELAY_TIME时直接显示内容，否则延迟显示骨架屏

EurFelux commented:
> > 给skeleton设置一个最小显示时间？这样不会闪，但总是要显示一小段时间
> 
> 把**延时显示**和**最小显示时间**都加上吧。
> 
> 延时显示，试试200ms。超过这个时间数据未好，则显示skeleton。 最小显示时间，试试500ms。skeleton一旦显示了，至少显示500ms。
> 
> 实际值要测试看看肉眼效果，目标是让人觉得自然而不闪

按你说的做了一版，感觉体验怪怪的。时间还得调整一下

EurFelux committed:
fix(messages): 修改显示逻辑

0xfullex commented:
> > > 给skeleton设置一个最小显示时间？这样不会闪，但总是要显示一小段时间
> > 
> > 
> > 把**延时显示**和**最小显示时间**都加上吧。
> > 延时显示，试试200ms。超过这个时间数据未好，则显示skeleton。 最小显示时间，试试500ms。skeleton一旦显示了，至少显示500ms。
> > 实际值要测试看看肉眼效果，目标是让人觉得自然而不闪
> 
> 按你说的做了一版，感觉体验怪怪的。时间还得调整一下

对，要肉眼细调一下

EurFelux committed:
perf(渲染): 调整messages渲染逻辑

调整skeleton显示时间与显示skeleton的延时；
调整渲染内容

EurFelux committed:
fix(Messages): 添加超时关闭skeleton的逻辑防止意外显示

EurFelux committed:
Merge branch 'main' into perf/router

EurFelux committed:
fix: 使用loggerService记录debug日志信息

EurFelux committed:
perf(渲染):为TopicsTab添加最小骨架屏显示时间

将骨架屏最小显示时间从300ms减少到150ms
移除首次渲染的特殊处理逻辑，统一使用骨架屏状态控制

Pleasurecruise commented:
是否要考虑基于简洁和气泡两种消息样式的骨架

EurFelux committed:
fix(FloatingSidebar): 修复浮动HomeTabs有双滚动条的问题和虚拟列表不加载的问题

将状态变量maxHeight重命名为height以更准确反映用途
移除click触发方式并调整悬浮延迟
修改PopoverContent样式为flex布局并使用height属性

EurFelux committed:
Merge branch 'main' into perf/router

EurFelux committed:
refactor(消息和话题组件): 优化骨架屏显示逻辑和日志级别

统一消息和话题组件的骨架屏显示逻辑，将日志级别从debug调整为silly
重构话题组件加载状态管理，增加过渡效果处理

EurFelux committed:
perf(Messages): 使用startTransition优化消息加载性能

将setTimeout替换为startTransition来优化消息加载的性能，避免不必要的延迟

EurFelux committed:
fix: 避免初始化时显示幻觉数据

EurFelux committed:
refactor(HomeTabs): 使用TabContainer组件优化标签页切换逻辑

将条件渲染改为通过TabContainer组件控制显示，这样不会触发tab的重新挂载，状态也不会在切换tab时重置

EurFelux committed:
fix: 调整骨架屏最小显示时间为200ms

EurFelux committed:
feat(话题标签页): 优化话题切换时的过渡效果

移除骨架屏加载逻辑，改为使用透明度过渡
添加willTransition状态控制过渡状态

EurFelux committed:
style: 注释掉 Messages.tsx 中部分日志输出

EurFelux committed:
Revert "refactor(HomeTabs): 使用TabContainer组件优化标签页切换逻辑"

This reverts commit 80f60a801d8aea0ee37f88968a2705753e85af49.

EurFelux committed:
feat: 使用 useDeferredValue 优化 TopicsTab 性能

在 TopicsTab 中，通过 useDeferredValue 对 displayedTopics 进行处理，
并将处理后的 deferredDisplayedTopics 应用于 DraggableList，
以优化性能。

EurFelux committed:
perf(AssistantsTab): 使用useDeferredValue优化助手列表渲染性能

EurFelux commented:
重做了TopicsTab的过渡效果，去除了skeleton，可以看看效果怎么样

https://github.com/user-attachments/assets/17268e79-ac3d-4937-84f3-41bc5f0b66bb

beyondkmp commented:
> 重做了TopicsTab的过渡效果，去除了skeleton，可以看看效果怎么样
> 
>  2025-07-24.01-00-48.mov

效果看起来不错

EurFelux committed:
Merge branch 'main' into perf/router

EurFelux commented:
> 冲突解决下？

done

beyondkmp commented:
有个问题，在一个肋手里面向上滚动到比较旧的消息了，然后切到其它助手，再切回来，这个时候就自动到最后的消息了，不会在之前滚动的位置那里。试了下生产的是会停留在之前的位置的。

EurFelux committed:
fix: 修复滚动位置记忆问题

EurFelux commented:
判断是由于初始渲染骨架屏触发不了滚动位置加载，现在让骨架屏消失后额外触发一次滚动位置加载。

beyondkmp commented:
> 判断是由于初始渲染骨架屏触发不了滚动位置加载，现在让骨架屏消失后额外触发一次滚动位置加载。

测试了下，修复了。

EurFelux commented:
> > 判断是由于初始渲染骨架屏触发不了滚动位置加载，现在让骨架屏消失后额外触发一次滚动位置加载。
> 
> 测试了下，修复了。但是还有问题，如果两个助手都向上滚动到历史消息，再去不断切换，就又全部会回到最后一条。好像只有一个助手滚动历史消息能正常

复现不出来，正式版上会这样吗

EurFelux committed:
refactor(useScrollPosition): 优化滚动位置处理并添加日志记录

移除不必要的useEffect，将滚动逻辑移至triggerScroll函数中
添加日志服务记录滚动位置和容器高度
使用更清晰的setTimeout语法

EurFelux committed:
fix(useScrollPosition): 避免运行使用值与日志打印不同步

beyondkmp commented:
我录屏给你看看，还有复现的数据也可以单独发你看看

alephpiece commented:
对话中的 skeleton 有点少，感觉很空...

alephpiece commented:
还有点问题：
- prompt 总是会先出现吗？这里会闪烁
- 有的话题似乎不会显示 skeleton，是阻塞了？

视频我切换到长话题的时候特别明显

https://github.com/user-attachments/assets/0e7f721f-0063-40b5-a280-47927720ffd5

EurFelux commented on src/renderer/src/pages/home/Tabs/TopicsTab.tsx:
```diff
@@ -423,26 +434,46 @@ const Topics: FC<Props> = ({ assistant: _assistant, activeTopic, setActiveTopic,
 
   const singlealone = topicPosition === 'right' && position === 'right'
 
+  const deferredDisplayedTopics = useDeferredValue(displayedTopics)
```
不是，是通过旧值和新值比较控制的样式变化
拖动我看看怎么能让它不闪吧

EurFelux committed:
refactor(ui): 移除过渡动画控制

EurFelux committed:
Merge branch 'main' into perf/router

EurFelux committed:
fix(useScrollPosition): 修复滚动位置恢复问题并优化触发逻辑

添加 timerRef 以清除未完成的定时器
将滚动逻辑提取到 triggerScroll 函数中
在 Messages 组件中根据 groupedMessages 变化触发滚动

EurFelux committed:
fix: 调整骨架屏最小显示时间为300ms

EurFelux committed:
feat(消息列表): 添加加载骨架屏并优化滚动行为

添加消息加载时的骨架屏效果，避免内容闪烁
优化滚动容器行为，默认滚动到底部
引入 SKELETON_MIN_TIME 常量控制最小加载时间

EurFelux committed:
fix(AssistantsTab): 不再使用deferredValue

EurFelux committed:
perf(AssistantsTab): 优化助手列表渲染性能

使用 startTransition 和状态管理来避免不必要的重新渲染

EurFelux committed:
Merge branch 'main' into perf/router

alephpiece commented:
滚动位置好像有点问题，在一个话题滚动到最顶上，切换到别的再切换回来，被滚动到底部去了

https://github.com/user-attachments/assets/15c12b5e-e977-486f-8bb3-35a829d881d7

EurFelux committed:
fix: 优化滚动控制

EurFelux committed:
refactor(hooks): 移动logger定义的位置

EurFelux committed:
fix: 恢复Messages组件中的调试日志输出

EurFelux committed:
fix(消息组件): 添加日志以调试显示消息时的加载状态

alephpiece commented:
> > 不知道是不是滚动触发的时机不对，只要话题里有很多 mermaid/plantuml 代码块这种延迟加载的，首次打开时，滚动条很有可能不在底部。
> 
> 滚动优化了一下
> 
> skeleton持续显示的问题我复现不出来，加了调试日志，能看看你那里什么情况吗

启动时：

<img width="531" height="228" alt="image" src="https://github.com/user-attachments/assets/6bf53e42-4b76-4840-803c-5b15c46c2007" />

再前面就是 app initialized 了，第一个助手的第一个话题是空的

EurFelux committed:
fix(消息列表): 修复消息滚动和布局问题

调整消息容器的滚动行为，将flex-direction改为column-reverse以保持新消息生成时滚动位置
将Prompt组件移动到InfiniteScroll下方并添加相关注释说明
仅在isLoaded为true时触发滚动

EurFelux committed:
refactor(Messages): 移除未使用的skeletonTimerCheckedRef并优化加载逻辑

简化消息加载逻辑，移除未使用的skeletonTimerCheckedRef变量
将setSkeletonTimerChecked移到更合理的位置
优化useEffect依赖项以提升性能

EurFelux committed:
Merge branch 'main' into perf/router

EurFelux committed:
style(Messages): 调整消息列表布局为反向列排列

alephpiece commented:
> 整个`Messages`回炉重造了，现在这个方案稳定很多。
> 
> `TopicsTab`的过渡效果不好控制，暂时去掉了。但是，保留了`startTransition`，依然能优化响应速度。
> 
> 给`AssistantsTab`也添加了延迟渲染。切换tab的响应速度变得非常快，缺点是延迟渲染导致助手列表在渲染出来之前会短暂地啥也没有。

可以在没有加载好的时候不要渲染那个“添加助手”，不然助手列表没出现的空挡会有个按钮，感觉会闪一下（看开发工具会比较明显）

EurFelux committed:
refactor(AssistantsTab): 优化拖拽列表更新逻辑并添加过渡状态

使用useTransition处理拖拽状态更新以避免阻塞UI
添加handleReorder函数统一处理列表顺序更新
移除不必要的注释和未使用的logger

EurFelux committed:
refactor(TopicsTab): 优化话题加载逻辑并使用startTransition

使用startTransition来优化话题加载的性能，避免阻塞UI渲染
将displayedTopics的初始状态改为undefined以明确加载状态

EurFelux committed:
Merge branch 'main' into perf/router

EurFelux committed:
Merge branch 'main' into perf/router
